File: frontend/src/lib/types/faq.ts
// FAQ document types for the frontend.

import type { Provider, SourceItem } from "./requests";

export interface FAQRequest {
  sources: SourceItem[];
  provider: Provider;
  model: string;
}

export interface FAQItem {
  id: string;
  question: string;
  answer: string;
  tags: string[];
}

export interface FAQMetadata {
  source_count: number;
  generated_at: string;
  tag_colors: Record<string, string>;
}

export interface FAQDocument {
  title: string;
  description?: string | null;
  items: FAQItem[];
  metadata: FAQMetadata;
}

export interface FAQProgressEvent {
  type: "progress";
  stage: string;
  percent: number;
  message?: string;
}

export interface FAQCompleteEvent {
  type: "complete";
  document: FAQDocument;
  download_url: string;
  file_path: string;
  session_id?: string | null;
}

export interface FAQErrorEvent {
  type: "error";
  message: string;
  code?: string;
}

export type FAQEvent = FAQProgressEvent | FAQCompleteEvent | FAQErrorEvent;

export function isFAQProgressEvent(event: FAQEvent): event is FAQProgressEvent {
  return event.type === "progress";
}

export function isFAQCompleteEvent(event: FAQEvent): event is FAQCompleteEvent {
  return event.type === "complete";
}

export function isFAQErrorEvent(event: FAQEvent): event is FAQErrorEvent {
  return event.type === "error";
}

export const TAG_GRADIENTS: Record<string, string> = {
  "blue-cyan": "from-blue-500 to-cyan-400",
  "purple-pink": "from-purple-500 to-pink-400",
  "orange-amber": "from-orange-500 to-amber-400",
  "green-teal": "from-emerald-500 to-teal-400",
  "rose-red": "from-rose-500 to-red-400",
  "indigo-violet": "from-indigo-500 to-violet-400",
  "yellow-lime": "from-yellow-400 to-lime-400",
  "slate-gray": "from-slate-500 to-gray-400",
};

File: backend/doc_generator/infrastructure/api/schemas/faq.py
"""Pydantic request and response models for FAQ generation."""

from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

from .requests import Provider, SourceItem


class FAQRequest(BaseModel):
    """Request model for FAQ generation."""

    sources: list[SourceItem] = Field(
        description="Single source (file, url, or text)",
        min_length=1,
        max_length=1,
    )
    provider: Provider = Provider.GEMINI
    model: str = "gemini-2.5-flash"

    model_config = ConfigDict(
        json_schema_extra={
            "examples": [
                {
                    "sources": [
                        {"type": "url", "url": "https://example.com/article"}
                    ],
                    "provider": "gemini",
                    "model": "gemini-2.5-flash",
                }
            ]
        }
    )


class FAQItemResponse(BaseModel):
    """A single FAQ item in the response."""

    id: str
    question: str
    answer: str
    tags: list[str] = Field(default_factory=list)


class FAQMetadataResponse(BaseModel):
    """Metadata in FAQ response."""

    source_count: int = 0
    generated_at: str
    tag_colors: dict[str, str] = Field(default_factory=dict)


class FAQDocumentResponse(BaseModel):
    """Complete FAQ document response."""

    title: str
    description: str | None = None
    items: list[FAQItemResponse] = Field(default_factory=list)
    metadata: FAQMetadataResponse


class FAQProgressEvent(BaseModel):
    """Progress event during FAQ generation."""

    type: Literal["progress"] = "progress"
    stage: str
    percent: float = Field(ge=0, le=100)
    message: str | None = None


class FAQCompleteEvent(BaseModel):
    """Completion event with FAQ data."""

    type: Literal["complete"] = "complete"
    document: FAQDocumentResponse
    download_url: str
    file_path: str
    session_id: str | None = None


class FAQErrorEvent(BaseModel):
    """Error event during FAQ generation."""

    type: Literal["error"] = "error"
    message: str
    code: str | None = None
